### ISSUE TITLE
Hide element for screen readers when disabled

### ISSUE THREAD

**Issue by giuseppeg** [danoc/clickable-box#11]
When `disabled` is `true` the element should get `aria-hidden="true"` so screen readers won't read its content.

**Comment by giuseppeg** [comment]
Btw sorry for opening issues instead of submitting PRs. They are more like reminders for me, I might contribute when I have some spare time ;)

**Comment by danoc** [comment]
No problem at all. I appreciate the thoughtful suggestions!

**Comment by scottaohara** [comment]
A disabled element is very different than one hidden to screen readers.

While a disabled element should not be accessible by use of the <kbd>tab</kbd> key, or perform its associated action on click/keypress, screen reader users SHOULD still be able to locate it by use of other navigational methods, e.g. the virtual cursor (up/down arrow keys in JAWS/NVDA).

Instead, being that these are not native `button`s, they should get an `aria-disabled="true"` set to them if they are disabled, and their `tabindex` should be set to -1.

```
<span
  // this becomes non-functional when disabled
  onClick={this.closeModal}
  // this gets a negative -1 to make it skipped by standard tabbing
  tabIndex={-1}
  // also non-functional when disabled
  onKeyDown={...}
  // Tell screen readers that the element is a button
  role="button"
  // update this to be the disabled cursor if you so wish
  style={{ cursor: 'pointer' }}
  // add ARIA attribute because native HTML attribute is not allowed on spans
  aria-disabled="true"
  // All other props are passed through to the element
  aria-label="Close modal"
  className="icon-button"
>
  <CloseIcon />
</span>
```

**Comment by giuseppeg** [comment]
ah cool thanks Scott! Is tabindex=-1 necessary if the element is a span without tabindex?

**Comment by scottaohara** [comment]
At a high level, being a `span` as the base element that's transformed into a "button", there'd be little theoretical difference between toggling the existence of the `tabindex` attribute to make this a focusable/non-focusable element.

Likely safer to just toggle the value of the `tabindex` though, as i can't be 100% positive on any adverse affects adding/removing the attribute might have.  In the end, it's slightly more effort to add/remove the attribute than it is to just change the value.

**Comment by giuseppeg** [comment]
Sure thing, thanks for the explanation. Another thing to consider is that `props.children` might contain focusable elements which should not be focusable when the button is `disabled`.

**Comment by scottaohara** [comment]
you mean children of the button?

a button should never have focusable children, regardless of state.

**Comment by giuseppeg** [comment]
I know but here is totally possible (easier?) also considering the library name "clickable box". Anyway that'd be people's fault in the end.

**Comment by danoc** [comment]
Thanks for the tips! Made two PRs to fix: #14 and #15.

**Comment by danoc** [comment]
Also added some accessibility best practices to the bottom of the README. 👍

**Comment by danoc** [comment]
Fixed in #14 and #15.

**Comment by giuseppeg** [comment]
awesome Daniel!! 

> disable pointer events with CSS

you might want to add this in a follow up PR

**Comment by danoc** [comment]
Thanks!

> disable pointer events with CSS

From what I understand this wouldn't make a difference since we remove the `onClick` if `disabled`. Is that correct?

**Comment by scottaohara** [comment]
those aren't really accurate test cases though because the onclick event should have been on the element that has the aria-disabled, as that would also be the element with role=button

your examples show the potential for nested interactive elements, which is again a failing on the developer to not realize they shouldn't be doing that.

**Comment by scottaohara** [comment]
e.g.
```
<!-- correct test case -->
<span role=button tabindex=-1 aria-disabled=true>no onClick </span>

<!-- this would be invalid and produce accessibility / general UX issues -->
<span onclick="whatever">
  <span role=button tabindex=-1 aria-disabled=true>
     im still disabled, but my parent has a click event for some reason?
  </span>
</span>
```

**Comment by giuseppeg** [comment]
event delegation is a quite common practice

**Comment by scottaohara** [comment]
i understand that. 

placing actual `onClick`s on parent elements to actionable elements causes issues for screen readers.  if an element has an onClick set to it, in the DOM, then it should be an actual interactive element

**Comment by danoc** [comment]
This seems to be a bit of a philosophical question about how closely `ClickableBox` should behave like a `button` if the `button` is being used incorrectly.

**Comment by giuseppeg** [comment]
I am not implying that it is a good practice but it happens. @danoc agreed, I just wanted to mention the issue. Up to you to decide what to do.

**Comment by scottaohara** [comment]
agreed @giuseppeg.

I'm just trying to point out things to keep in line with:
> "without sacrificing accessibility"

people are going to misuse stuff, and there's not much that can be done about that.  one of the hardest parts about recreating HTML elements is keeping them as close to native functionality as you can, potential warts n' all.
